### PR TITLE
Add support for HSV encoded color lights

### PR DESCRIPTION
**Pull request by postlund**
This PR adds support for light bulbs with colors encoded in HSV format. Brightness is supported as well (since it's embedded in the color data).

@rospogrigio This PR is ready for review, but it should not be merged before #92 and a final test by user.

### CODE REVIEW

**Comment by SmartM-ui** [comment]
Here I am,
I follow the development with interest.
As soon as possible I try and send the log.
If I can be useful for anything, don't hesitate to ask

**Comment by postlund** [comment]
Would be great with a tester for this 👍 Do you have bulbs configured via RGB as well?

**Comment by ultratoto14** [comment]
Hi i do have multiple lights with different settings:

- HSV on 4 bytes (0-1000) range
- RGB_HSV (0-255) range

I have something almost fully working here [https://github.com/ultratoto14/localtuya-homeassistant/tree/rgbw_update](https://github.com/ultratoto14/localtuya-homeassistant/tree/rgbw_update). Be kind, i'm not a python developer, many things can be done differently 
You can have a check to implement and merge what you need in yours.
As a high number of DP is used, i tried to suggest one for each entry in case i can found it during the config

What is not working in mine:

- I have RGBW GU10 from different manufacturer which have different white range. So mired should be also configurable
- I need to detect when HA sends white or color commands to adapt, as it is separated. When a scene is saved via the editor it stores all the information, meaning brightness, color_temp, color and everything is sent to the bulb once the scene is triggered. Need to manually update the yaml to ensure only the needed command is sent.

**Comment by postlund** [comment]
Thanks @ultratoto14, I will a greater look. I think I have implemented most things in a similar manner, not RGB though. But I can take some inspiration from your code for that. One change I'm planning is to use a single option for color values, as HSV or RGB can be deduced from length of the value. 

Color temperature is a different matter, related of course. Should be simple enough to implement (there is also an issue about it: #91). I intend to add two options to allow custom values for lower and upper range, specified in Kelvin and converted to mired internally.

**Comment by ultratoto14** [comment]
I used the set_dps you introduced, i fixed the unneeded value and put the corresponding in common.py

My full list of devices:

- Led Strip HSV mode
- GU10 RGBW RGB mode
- GU10 RGBW HSV mode
- Color Bulb HSV mode

They are from different manufacturers and all the one in HSV are using this 1000 range where the RGB_HSV are using the 255 range. 
Do you know HSV lights using the 255 range ?

**Comment by postlund** [comment]
I would not expect HSV to be limited to 255 like that. This picture explains HSV pretty well:

**Comment by ultratoto14** [comment]
;-) my question was not clear. Do you know tuya lights that uses HSV coding for the color and 255 range for brightness and color_temp ? Because if you deduce the kind of color from the length, you can then also deduce the max value for brightness and color_temp.
As many things should be setup for a simple light, if the configurator can help the user in choosing the configuration values, it's a big addition.

**Comment by postlund** [comment]
Ah, yes, then I understand. No, I don't know about that and I would expect that to not exist either. They however seem to use the same brightness range for white and color, so no distinction needs to be made there. Color temperature uses a different range based on mirad/kelvin range of the lamp and has to be entered separately.

**Comment by postlund** [comment]
Tested and only needs review before merging!

**Comment by ultratoto14** [comment]
Just tested it, two main remarks:

- Color temp range is not aligned but perhaps it's not the purpose of this PR
- When going from colour to white mode, the bulb icon do not switch to the white and keep the old color setup

**Comment by postlund** [comment]
@ultratoto14 Thanks for testing! Color temperature will be dealt with in #91. Regarding the icon, not sure how that works. Will have to look into that, thanks for reporting!

**Comment by postlund** [comment]
@ultratoto14 Pushed an update now, can you try if icon works as expected?

**Comment by ultratoto14** [comment]
@postlund Icon is now working as expected.

**Comment by postlund** [comment]
@ultratoto14 Brilliant, thanks! 🎉

**Comment by ultratoto14** [comment]
@postlund, i found a remaining small bug about the consistency of the brightness setting.
If you go from white to colour, the brightness is conserved. I think it's the desired behavior.
Once in color, i you change the brightness, it's still working. But when you go back to white, the brightness is the same as before going to color.
Steps to reproduce:

- Put in white - OK
- Set brightness to 50% - OK
- Choose a color - OK brightness still at 50%, color is OK too
- Change brightness to 100% - OK
- Change color_temp to get back to white: KO brightness is back to 50%

**Comment by postlund** [comment]
@ultratoto14 Ok, that's interesting. Maybe I should set white and color brightness at the same time to keep them in sync. I guess we could try that.

**Comment by postlund** [comment]
@ultratoto14 Loong-shot here, but please try the latest commit and see if it works as expected (or at all).

**Comment by ultratoto14** [comment]
Hi @postlund, sorry for the delay. I still have a couple of hours to test before being unavailable for 3 days.
Unfortunately, the fix does not work. During the test , at the third step, changing brightness when in color, the bulb witch back to white.
Setting the brightness dp when in color mode is switching the bulb to white, it should not be set every time.

**Comment by lapy** [comment]
Hi guys,

I made a change all by myself (I hadn't noticed this pull request). Here is the commit that I made on my fork:

**Comment by lapy** [comment]
Hi guys,

I hadn't noticed this pull request so I added this feature by myself on my fork.
I tested it on a tuya compatible E14 bulb.

Working:
- Color selection
- Color brightness
- Improved white temp mapping (for me the original code only covered 1/4 of the range)

Please take a look if that helps! (another set of eyes/ideas).

https://github.com/lapy/localtuya/commit/7d76d915afae661c862ce32bde9aa1010c74f2ba
Please ignore the workspace.xml file :)

Thank you,
Vince

**Comment by ultratoto14** [comment]
Hi @postlund, tried to find the problem and it seems that the update and turn on are competing.
You can update the brightness when changing the color_temp between line 171 and 172.
```python
          if not self._is_white_mode:
            await self._device.set_dp(self._brightness, self._config.get(CONF_BRIGHTNESS))
            await self._device.set_dp("white", self._config.get(CONF_COLOR_MODE))
```
You need to apply the brightness first because of the status updated.
When you will use a single call to apply multiple DPs, the problem may be solved.

**Comment by postlund** [comment]
@ultratoto14 Seems like @lapy has come a longer way than I have (especially since I haven't prioritized this recently), maybe you can try his commit and see it if works better? At the end of the day I would love to see someone else implement this so I can focus on core stuff instead 🙃

**Comment by postlund** [comment]
@lapy Send a PR, I can review it for inclusion here and @ultratoto14 can probably test it.

**Comment by SmartM-ui** [comment]
> @ultratoto14 Seems like @lapy has come a longer way than I have (especially since I haven't prioritized this recently), maybe you can try his commit and see it if works better? At the end of the day I would love to see someone else implement this so I can focus on core stuff instead 🙃

Hi @postlund ,
if I can help I am available to carry out some tests.

Which version are you testing?

**Comment by lapy** [comment]
Hi,
This is my pull request https://github.com/rospogrigio/localtuya/pull/115 if you wanna test it.

Thank you,
Vince

**Comment by SmartM-ui** [comment]
> Hi,
> This is my pull request #115 if you wanna test it.
> 
> Thank you,
> Vince

OK! Thanks

**Comment by SmartM-ui** [comment]
Hi @postlund ,
I tried the pull # 115 and I must say that as far as the color temperature is concerned it manages to align correctly with the smartlife app.

I was also testing this version # 90, but having configured the bulbs through the yaml file, I don't know what code I need to enter in order to manage the colors too.

Could you kindly tell me how to set yaml file?

Thanks

**Comment by postlund** [comment]
@ultratoto14 Got your email, will respond here so we can decide on how to move forward.

From my point-of-view, it doesn't really matter if we go with my solution or the one by @lapy. I do feel that it'a a bit problematic to not have a steady maintainer behind the code we merge. So I need your input, based on what we have. Should we merge this and continue adding RGB support from @ultratoto14 as well as what else is missing in additional steps? Or should we go with @lapy? I'm up for anything, we just need to increase the pace a bit and get something merged that we can extend upon.

**Comment by ultratoto14** [comment]
@postlund, I totally agree with you. I just want to have this included to ensure i can work with it without implementing RGB each time because it's not yet supported. @lapy said he returned his bulb, so he will not maintain the code.
The branch I posted just fixed the brightness pb of your PR, add RGB bulb support and fixed the color_temp. 
If it's Ok, I open a PR. As it is based on this one, you close it and link to mine.

**Comment by postlund** [comment]
Sounds good to me 👍 Let's merge this and continue with additional PRs and writing issues for bugs.

@rospogrigio You can review this now.

File: custom_components/localtuya/common.py
Comment (rospogrigio) [inline custom_components/localtuya/common.py:222]: Do we really want to keep this debug message?

**Review by rospogrigio — APPROVED** [review]
Just one comment, for the rest I think it's ok.

File: custom_components/localtuya/common.py
Comment (postlund) [reply custom_components/localtuya/common.py:222]: I do like this debug print as it will tell us if a certain option is configured or not (it helps when a user thinks a he/she configured an option, but actually didn't).

**Comment by postlund** [comment]
@ultratoto14 This is merged now. You can rebase your changes on top of `master` and open a new PR with your changes now.

**Comment by SmartM-ui** [comment]
@postlund 
Hi Pierre,
to test the new job which pull should i follow?

**Comment by ultratoto14** [comment]
@postlund, @rospogrigio, @SmartM-ui the [PR120](https://github.com/rospogrigio/localtuya/pull/120) is created

**Comment by SmartM-ui** [comment]
> @postlund, @rospogrigio, @SmartM-ui the [PR120](https://github.com/rospogrigio/localtuya/pull/120) is created

Thanks, I'll try now

**Pull request merged**
